x-pack/plugins/code/server/routes/redirect_socket.ts

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { Server } from 'hapi';
+import SocketIO, { Socket } from 'socket.io';
+import ClientIO from 'socket.io-client';
+
+import { SocketKind } from '../../model';
+import { Log } from '../log';
+import { ServerOptions } from '../server_options';
+import { BASE_PLACEHOLDER, mainNodeBaseUrl } from './redirect';
+
+export async function redirectSocketRoute(server: Server, serverOptions: ServerOptions, log: Log) {
+  const socketIO = SocketIO(server.listener, { path: '/ws' });
+  let redirectUrl = serverOptions.redirectToNode;
+  const hasBaseUrl = redirectUrl.includes(BASE_PLACEHOLDER);
+
+  async function connectToMainNode() {
+    let baseUrl = '';
+    if (hasBaseUrl) {
+      baseUrl = await mainNodeBaseUrl(redirectUrl);
+      redirectUrl = serverOptions.redirectToNode.replace(BASE_PLACEHOLDER, '');
+    }
+    const socketUrl = `${redirectUrl}`;
+    return ClientIO(socketUrl, { path: `${baseUrl}/ws` });
+  }
+  // maintain a connection to main node
+  let mainNodeSocket = await connectToMainNode();
```

Comment:
This socket needs authentication too

Reply:
See the comment [here](https://github.com/elastic/kibana/pull/27012#discussion_r249771016). Also, the socket is used mostly as an internal long-lived connection between non-code node and code node just for message forwarding across kibana servers.

Reply:
We'll still need to authenticate the socket connection though, even if it's only used server -> server

Reply:
Just created another PR https://github.com/elastic/kibana/pull/29246 to apply the same socket authentication of client -> codeNode connection to the client -> nonCodeNode in current HA setup. This is to make sure that the initial socket connection is secured. The long-lived nonCodeNode server -> codeNode server connection with be created with a reserved user 'code_user'.